src/flb_utils.c

```diff
@@ -744,23 +746,38 @@ int flb_utils_write_str(char *buf, int *off, size_t size,
 
             is_valid = FLB_TRUE;
             for (b = 0; b < hex_bytes; b++) {
-                /* Trailing characters must start with bit 1 */
-                if (b > 0 && !(str[i+b] & 0x80)) {
-                    /* Invalid unicode character. skip */
-                    flb_warn("[pack] unexpected UTF-8 continuation byte, "
-                             "omitting a character");
+                /* Leading characters must start with bits 11 */
+                if (b == 0 && ((str[i] & 0xC0) != 0xC0)) {
+                    /* Invalid unicode character. replace */
+                    flb_debug("[pack] unexpected UTF-8 leading byte, "
+                             "substituting character with replacement character");
+                    ++i; /* Consume invalid leading byte */
+                    is_valid = FLB_FALSE;
+                    break;
+                }
+                /* Trailing characters must start with bits 10 */
+                else if (b > 0 && ((str[i] & 0xC0) != 0x80)) {
+                    /* Invalid unicode character. replace */
+                    flb_debug("[pack] unexpected UTF-8 continuation byte, "
+                             "substituting character with replacement character");
+                    /* This byte, i, is the start of the next unicode character */
                     is_valid = FLB_FALSE;
                     break;
                 }
 
-                tmp[b] = str[i+b];
+                tmp[b] = str[i];
+                ++i;
             }
+            --i;
 
             if (is_valid) {
                 encoded_to_buf(p, tmp, hex_bytes);
                 p += hex_bytes;
             }
-            i += (hex_bytes - 1);
+            else {
+                encoded_to_buf(p, flb_utils_replacement_character, 3);
```

Comment:
annoying improvement suggestion : wouldn't it be nice if instead of replacing the broken sequence we escaped those borked code units? 

`"\xe3\x81\x01abc"` would become `"\\xe3\\x81\\x01abc"` instead of `"�\\u0001abc"` which in a way would be more representative.

We could add the � to mark the broken sequence too if we document it, it just seems less lossy.

Reply:
I did consider this but \\xe3 is an invalid JSON sequence, and this function is already used in several places (msgpack and cloudwatch api) to sanitize text for JSON. [ \\\\xe3 could overlap with unescaped text and should also be avoided ]

I also considered:
- escaping \xe3\x81\x01abc to something like \\u00e3\\u0081\\u0001
however I'm pretty sure that the codepoints such as \\u0081 will overlap with ascii codepoints see: https://en.wikipedia.org/wiki/UTF-8#Encoding and thus should be avoided
- just writing the hex e38101, but that would overlap with generic text
- leaving the invalid fragments, but that would potentially mean the string is not safe/escaped
- deleting the invalid fragments, which is fine, but adding the replacement character seems like an option that gives more information.

After meeting with some people on the team, we determined that the replacement character would most likely be the best way to deal with the invalid fragments. What do you think?

Reply:
@leonardo-albertovich

Reply:
Considering that \x## could break JSON blocks we could use something like that ? sign as a delimiter and put the list of hex digits between them. 

Something like this : `� e3 81 01 �abc`

What do you think of that? The idea would be not losing data that might help the person affected by the issue figure it out.

Reply:
� e3 81 01 �abc is actually valid Unescaped Unicode so a user could actually have a log with � e3 81 01 �abc. It would be impossible to distinguish this from the invalid fragment once the text is escaped. I think ambiguous data is worse than lost data.

If you want this solution, I can work on it. Just please confirm that you are considering that the output data is ambiguous.

Reply:
I guess in the same sense a single � would be valid. Maybe then we could add a disclaimer such as "� corrupted utf-8 sequence : e3 81 01 �"

Now that's quite the footprint for an error but if that's possible then I'd rather go with that (or a shorter variation) than discarding data.

Reply:
Great. I'll work on a fix and post it to this pr hopefully by today.

Reply:
Awesome!

Reply:
done.